substrate/frame/support/procedural/src/pallet/parse/view_functions.rs

```
@@ -0,0 +1,144 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governsing permissions and
+// limitations under the License.
+
+use frame_support_procedural_tools::get_doc_literals;
+use inflector::Inflector;
+use syn::spanned::Spanned;
+
+/// Parsed representation of an impl block annotated with `pallet::view_functions_experimental`.
+pub struct ViewFunctionsImplDef {
+	/// The where_clause used.
+	pub where_clause: Option<syn::WhereClause>,
+	/// The span of the pallet::view_functions_experimental attribute.
+	pub attr_span: proc_macro2::Span,
+	/// Docs, specified on the impl Block.
+	pub docs: Vec<syn::Expr>,
+	/// The view function definitions.
+	pub view_functions: Vec<ViewFunctionDef>,
+}
+
+impl ViewFunctionsImplDef {
+	pub fn try_from(attr_span: proc_macro2::Span, item: &mut syn::Item) -> syn::Result<Self> {
+		let item_impl = if let syn::Item::Impl(item) = item {
+			item
+		} else {
+			return Err(syn::Error::new(
+				item.span(),
+				"Invalid pallet::view_functions_experimental, expected item impl",
+			))
+		};
+		let mut view_functions = Vec::new();
+		for item in &mut item_impl.items {
+			if let syn::ImplItem::Fn(method) = item {
```

we can also return an error for not `Fn` items.

---

Done in https://github.com/paritytech/polkadot-sdk/pull/4722/commits/de5834ca82d064f8ae66958a732cc00a03de704b